plutus-metatheory/src/Cost/Model.lagda.md

```diff
@@ -0,0 +1,194 @@
+runModel (multipliedSizes i s) xs = i + s * (prod xs)
+runModel (minSize i s) xs = i + s * minimum xs
+runModel (maxSize i s) xs = i + s * maximum xs
+runModel (twoArgumentsLinearInXAndY i s₁ s₂) (a ∷ b ∷ []) = i + s₁ * a + s₂ * b 
+runModel (twoArgumentsSubtractedSizes i s min) (a ∷ b ∷ []) = i + s * (min ⊔ (a ∸ b))
+runModel (twoArgumentsLinearOnDiagonal c i s) (a ∷ b ∷ []) = 
+    if a ≡ᵇ b 
+      then i + s * a 
+      else c
+runModel (twoArgumentsConstAboveDiagonal c m) (a ∷ b ∷ []) = 
+    if a <ᵇ b 
+      then c 
+      else runModel m (a ∷ b ∷ [])
+runModel (twoArgumentsConstBelowDiagonal c m) (a ∷ b ∷ []) =
+    if b <ᵇ a 
+      then c 
+      else runModel m (a ∷ b ∷ [])
+```
+
+## Convert from Raw Model
+
+May fail if the model doesn't correspond to the number of arguments.
+
+```
+convertRawModel : ∀{n} → RawModel → Maybe (CostingModel n) 
+convertRawModel (ConstantCost c) = just (constantCost c)
+convertRawModel (AddedSizes (mkLF intercept slope)) = just (addedSizes intercept slope)
+convertRawModel (MultipliedSizes (mkLF intercept slope)) = just (multipliedSizes intercept slope)
+convertRawModel {suc n} (MinSize (mkLF intercept slope)) = just (minSize intercept slope)
+convertRawModel {suc n} (MaxSize (mkLF intercept slope)) = just (maxSize intercept slope)
+convertRawModel {suc n} (LinearCost (mkLF intercept slope)) = just (linearCostIn zero intercept slope)
+convertRawModel {suc n} (LinearInX (mkLF intercept slope)) = just (linearCostIn zero intercept slope)
+convertRawModel {suc (suc n)} (LinearInY (mkLF intercept slope)) = just (linearCostIn (suc zero) intercept slope)
+convertRawModel {suc (suc (suc n))}(LinearInZ (mkLF intercept slope)) = just (linearCostIn (suc (suc zero)) intercept slope)
+convertRawModel {2} (SubtractedSizes (mkLF intercept slope) c) = just (twoArgumentsSubtractedSizes intercept slope c)
+convertRawModel {2} (ConstAboveDiagonal c m) = map (twoArgumentsConstAboveDiagonal c) (convertRawModel m)
+convertRawModel {2} (ConstBelowDiagonal c m) = map (twoArgumentsConstBelowDiagonal c) (convertRawModel m)
+convertRawModel {2} (LinearOnDiagonal (mkLF intercept slope) c) = just (twoArgumentsLinearOnDiagonal c intercept slope)
+convertRawModel _ = nothing
+
+convertCpuAndMemoryModel : ∀{n} → CpuAndMemoryModel → Maybe (BuiltinModel n)
+convertCpuAndMemoryModel (mkCpuAndMemoryModel cpuModel memoryModel) with convertRawModel cpuModel | convertRawModel memoryModel 
+... | just cm | just mm = just (record { costingCPU = cm ; costingMem = mm })
+... | just _ | nothing = nothing
+... | nothing | m = nothing
+```
+
+## Creation of mapping function 
+
+Creates a function mapping builtins to their corresponding costing models, 
+starting from a `BuiltinCostMap`.
+
+We need to construct a `BuiltinModel (arity b)` for each `b`, and this may fail if
+the model int the map doesn't correspond to the arity. 
+
+```
+getModel : Builtin → BuiltinCostMap → Maybe (Σ Builtin (λ b → (BuiltinModel (arity b))))
```

Comment:
It works by trying to obtain a model for every `Builtin` in the meta-theory. So, if in the JSON file there is a cost model for a new function that has not been added to the metatheory it will simply be ignored. As you say, I think this is good, as usually new functions are added to the production implementation first. 

(BTW, at first I've got confused by your comment because the function returns a `Maybe` with a `nothing` constructor)